compiler/luci/pass/src/FoldDequantizePass.cpp

```diff
@@ -0,0 +1,190 @@
+      if (qd >= const_node->quantparam()->zerop.size())
+        qd = 0;
+
+      new_const_node->at<loco::DataType::FLOAT32>(i) =
+          (float)(const_node->at<loco::DataType::S8>(i) - const_node->quantparam()->zerop.at(qd)) *
+          const_node->quantparam()->scale.at(qd);
+    }
+  }
+  else
+  {
+    for (uint32_t i = 0; i < const_node->size<loco::DataType::U8>(); ++i)
+    {
+      uint32_t qd = (i % right_count) / (right_count / q_dim_value);
+      if (qd >= const_node->quantparam()->zerop.size())
+        qd = 0;
+
+      new_const_node->at<loco::DataType::FLOAT32>(i) =
+          (float)((int)const_node->at<loco::DataType::U8>(i) -
+                  const_node->quantparam()->zerop.at(qd)) *
+          const_node->quantparam()->scale.at(qd);
+    }
+  }
+
+  return new_const_node;
+}
+
+} // namespace
+
+namespace luci
+{
+
+/**
+ * Folding pattern 1 - When input of Dequantize is constant
+ *
+ * [Before]
+ *     quantized_const_input ----- Dequantize ----- Op1 ---
+ *                             +-- Op2 ---
+ *                             +-- Op3 ---
+ *
+ * [After]
+ *   dequantized_const_input ---------------------- Op1 ---
+ *
+ *     quantized_const_input ----- Op2 ---
+ *                             +-- Op3 ---
+ *
+ *
+ * Folding pattern 2 - When input of Dequantize is using quantized value
+ *
+ * [Before]
+ *     quantized_const_input ----- Gather ------ Dequantize --- Op1 ---
+ *                             +-- Op2 ---
+ *                             +-- Op3 ---
+ *
+ * [After]
+ *   dequantized_const_input ------Gather ------ Op1 ---
+ *
+ *     quantized_const_input ----- Op2 ---
+ *                             +-- Op3 ---
+ *
+ *
+ */
+bool FoldDequantizePass::run(loco::Graph *g)
+{
+  bool changed = false;
+
+  for (auto node : loco::all_nodes(g))
+  {
+    if (auto circle_dequant = dynamic_cast<luci::CircleDequantize *>(node))
+    {
+      auto input_node = circle_dequant->input();
+      if (auto const_input = dynamic_cast<luci::CircleConst *>(input_node))
+      {
+        // Input of Dequantize is constant
+        loco::replace(circle_dequant).with(dequantized_const_node(const_input));
+        changed = true;
+      }
+      else if (loco::dtype_get(input_node) == loco::DataType::FLOAT32)
+      {
+        // Input is already dequantized
+        loco::replace(circle_dequant).with(input_node);
+        changed = true;
+      }
+    }
+    else if (auto const_node = dynamic_cast<luci::CircleConst *>(node))
+    {
+      if (const_node->quantparam() != nullptr && is_foldable_const(const_node))
+      {
+        for (auto s : loco::succs(const_node))
+        {
+          auto s_user = loco::succs(s);
+          if (!is_hybrid_kernel_supported(s) && s_user.size() == 1 &&
+              dynamic_cast<luci::CircleDequantize *>(*s_user.begin()) != nullptr)
+          {
+            // Input can be pre-dequantized
+            changed |= replace_const_node(s, dequantized_const_node(const_node));
```

Comment:
It doesn't need to change the dtype of `s`? I mean change from `U8` or `S8` to `F32`.

Reply:
Dequantize is changing U8/S8 to F32... what do you mean @.@?

Reply:
`s` has got U8/S8 dtype first. After `replace_const_node` function is called, `s` has F32 params. It is internally applied somewhere?

Reply:
Ah, I understood it now. If `s` is replaced, it means that `FoldDequantize` cause some changes.
Then `TypeInfrencePass` will be done at `CircleOptimizer`.
In that phase, the inferred type will be changed :)

Reply:
> In that phase, the inferred type will be changed :)

Why not change the dtype of CircleNode as well? Is there a reason to keep the previous dtype of `s`?

Mismatch between the inferred type and the dtype of CircleNode seems risky. For example, let's assume another pass needs to use the dtype of `s`. Which type the pass should use? inferred type (fp32) or dtype saved in CircleNode (u8 or s8) ?

Reply:
> Mismatch between the inferred type and the dtype of CircleNode seems risky

Yes, it is very risky. It should be changed in view of long term.

> Which type the pass should use? inferred type (fp32) or dtype saved in CircleNode (u8 or s8) ?

**Inferred type**. During optimizations or transformations, new nodes can be generated. But if we decide to use saved dtype in CircleNode, we should always inference the shape and dtype for them. Furthermore, `ShapeInferencePass` and `TypeInferencePass` are implemented to do exactly that thing.

Reply:
> It will be done at TypeInferencePass.

I can't find the code that changes the dtype of CircleNode in TypeInferencePass. Can you tell me the location of the code?

Even though TypeInferencePass changes the dtype, it is safe to change the dtype here, because other passes can access the dtype of `s` before TypeInferencePass is called. This may cause a bug very difficult to find.

> Inferred type. During optimizations or transformations, new nodes can be generated. But if we decide to use saved dtype in CircleNode, we should always inference the shape and dtype for them. Furthermore, ShapeInferencePass and TypeInferencePass are implemented to do exactly that thing.

I didn't know that. There is only one pass that uses `dtype_get` (ResolveCustomOpMatMulPass). Most passes use dtype() in CircleNode. As I said above, it would be safe to set the correct dtype inside an optimization pass to avoid the type-mismatch bugs.

Reply:
> I can't find the code that changes the dtype of CircleNode in TypeInferencePass. Can you tell me the location of the code?

My words causes misunderstanding, sorry. `TypeInferencePass` do not change the dtype of CircleNode but change loco dtype.

> Even though TypeInferencePass changes the dtype, it is safe to change the dtype here, because other passes can access the dtype of s before TypeInferencePass is called. This may cause a bug very difficult to find.

Hmm... I agree with your concern but do not agree with the solution.
Until now, fortunately, those errors are not found but it can be occurred for later. So we should resolve this issue.
However, updating shape and dtype in every pass seems not good because

- Inferencing shape and dtype in each pass requires implementation of shape and dtype inference code. It can be crashed with the logic in `CircleTypeInferenceRule.cpp` and `CircleShapeInferenceRule.cpp`. In short, if some of shape is wrong, we don't know where is wrong and if some rule of shape/dtype inference is changed, we should change every pass related with it.
- If some pass generate many nodes, for example 10 nodes, implementation for all of them is really tough.

So I suggest to run `ShapeInferencePass` and `TypeInferencePass` after each pass ended at `CircleOptimizer`.
If then, when each passes ended, all of nodes will have updated shape and dtype.

> Most passes use dtype() in CircleNode. 

Yes, `FuseBCQPass` which I wrote do not use `loco::shape_get` and `loco::dtype_get` neither :(
I think that we should fix those things in every pass.

Reply:
> However, updating shape and dtype in every pass seems not good because

It seems that there was a misunderstanding. What I suggested was just adding one line of code in this pass.

`s->dtype(const_node->dtype());`

Adding one more line for assertion would be better.

`assert(const_node->dtype() == loco::DataType::Float32);`

> So I suggest to run ShapeInferencePass and TypeInferencePass after each pass ended at CircleOptimizer.
If then, when each passes ended, all of nodes will have updated shape and dtype.

This is not a complete solution, because type/shape are not updated until each pass ends. For example, let's assume a pass changes a type of a node at the beginning. The pass can access the type of the node again (before exiting). In this case, TypeInferencePass is not called yet (because the pass is not finished), so the type is not updated.

I think we had better use dtype instead of dtype_get inside a pass. And, dtype of CircleNode should be correctly updated whenever changes occur.

Reply:
Result of offline discussion is as following
- For now, insert dtype setting code
- However, the fundamental issues should be handled in the near future.
    - Risk of redundant shape/dtype information among CircleNode and locoNode
    - shape/dtype validity before and after each pass
    - Recommendation method for getting shape/dtype in each pass